### ISSUE TITLE
x-forwarded-proto set incorrect on outgoing API calls

### ISSUE THREAD

**Issue by gganssauge** [Haufe-Lexware/wicked.haufe.io#23]
My API is returning links with wrong schema if used from the API portal.
When using the API directly the links are returned with the correct scheme.
The reason is the setting of the request header _X-Forwarded-Proto_ which is always _http_ even when the wicked portal is only serving https.

Interestingly the value of this particular header is displayed as _https_ in the kong http-log, but it arrives as _http_ at my API server.

**Comment by gganssauge** [comment]
This is the request:
`curl -X GET --header 'Accept: application/hal+json' --header 'X-ApiKey: *************************************************' 'https://api.test.smartpilot.haufe.io/bookmarks-api/v1/'`

This is the http-log output:
`portal-logger_1            | {"started_at":1476983736484,"response":{"status":200,"size":"742","headers":{"server":"nginx/1.11.3","content-type":"application/hal+json","connection":"close","strict-transport-security":"max-age=31536000","access-control-allow-origin":"*","x-kong-proxy-latency":"0","x-kong-upstream-latency":"30","vary":"Origin","via":"kong/0.8.3","x-powered-by":"Servlet/3.0 JSP/2.2 (GlassFish Server Open Source Edition 3.1.2.2 Java/Oracle Corporation/1.7)","transfer-encoding":"chunked"}},"authenticated_entity":{"consumer_id":"4a5222e2-4bf6-4031-b8ca-bfa8f40d707c","id":"572cdafe-b733-424e-9c46-37cdc05012cb"},"request":{"method":"GET","uri":"/bookmarks-api/v1/","size":"258","request_uri":"http://api.test.smartpilot.haufe.io:8000/bookmarks-api/v1/","querystring":{},"headers":{"x-consumer-id":"4a5222e2-4bf6-4031-b8ca-bfa8f40d707c","x-forwarded-for":"91.66.22.165","x-consumer-groups":"bookmarkapi","accept":"application/hal+json","correlation-id":"61237abf-e82c-4454-941c-d7dbe7c6afe6","x-consumer-username":"test$bookmarkapi","host":"api.test.smartpilot.haufe.io","x-forwarded-proto":"https","x-forwarded-port":"443","authorization":"Basic ********************************************************","x-consumer-custom-id":"e3f6ac2bbefc2331bfc7b839ab7c55a8d8df3251","user-agent":"curl/7.35.0"}},"client_ip":"91.66.22.165","api":{"upstream_url":"https://smartpilot.haufe.io","request_path":"/bookmarks-api/v1","id":"9322d94b-de29-4376-9aa8-be958f9888d4","name":"bookmarkapi","preserve_host":true,"created_at":1476811771000,"strip_request_path":false},"latencies":{"request":32,"kong":2,"proxy":30}}
`

and finally my access log
`sslproxy     | nginx.1    | smartpilot.haufe.io 91.66.22.165 - api-user [20/Oct/2016:17:00:14 +0000] "GET /bookmarks-api/v1 HTTP/1.1" 200 280 "-" "curl/7.47.0" "-" "JS summary\x0A\x0AMethod: GET\x0AHTTP version: 1.1\x0AHost: smartpilot.haufe.io\x0ARemote Address: 91.66.22.165\x0AURI: /bookmarks-api/v1\x0AHeaders:\x0A  header 'Host' is 'smartpilot.haufe.io'\x0A  header 'Authorization' is 'Basic YXBpLXVzZXI6YVpwdGFmM0luU3RBQU1hb21nTzA='\x0A  header 'User-Agent' is 'curl/7.47.0'\x0A  header 'Accept' is 'application/hal+json'\x0AArgs:\x0A"`

**Comment by gganssauge** [comment]
Finally the output (note the wrong schema in the links):
`{"_links":{"self":{"href":"http://api.test.smartpilot.haufe.io/bookmarks-api/v1/"},"start":[{"href":"http://api.test.smartpilot.haufe.io/bookmarks-api/v1/"}]},"collections":{"user":"http://api.test.smartpilot.haufe.io/bookmarks-api/v1/user/{hmgUserId}"},"name":"bookmark-api","version":"1.0"}`

When I use the API directly I get (correct schema in the links):
`{"_links":{"self":{"href":"https://smartpilot.haufe.io/bookmarks-api/v1/"},"start":[{"href":"https://smartpilot.haufe.io/bookmarks-api/v1/"}]},"collections":{"user":"https://smartpilot.haufe.io/bookmarks-api/v1/user/{hmgUserId}"},"name":"bookmark-api","version":"1.0"}`

**Comment by gganssauge** [comment]
my API is configured as follows:
<pre>{
  "api": {
    "name": "bookmarkapi",
    "upstream_url": "$PORTAL_APIS_BOOKMARKAPI_UPSTREAM_URL",
    "request_path": "/bookmarks-api/v1",
    "preserve_host": true
  },
  "plugins": [
    {
      "config": {
        "origin": "*",
        "methods": "GET,HEAD,PUT,PATCH,POST,DELETE"
      },
      "name": "cors"
    },
    {
      "config": {
        "path": "/usr/local/kong/logs/kong-access.log"
      },
      "name": "file-log"
    },
    {
      "config": {
        "header_name": "Correlation-Id",
        "generator": "uuid"
      },
      "name": "correlation-id"
    },
    {
      "name": "http-log",
      "config": {
        "http_endpoint": "${PORTAL_LOGGING_URL}",
        "method": "POST",
        "timeout": "1000",
        "keepalive": "1000"
      }
    },
    {
      "name": "request-transformer",
      "config": {
        "add": {
          "headers": [
            "$PORTAL_APIS_BOOKMARKAPI_BASICAUTH"
          ]
        }
      }
    }
  ]
}</pre>

**Comment by DonMartin76** [comment]
In your case, change the `request-transformer` plugin configuration like this:

```json
{
      "name": "request-transformer",
      "config": {
        "add": {
          "headers": [
            "$PORTAL_APIS_BOOKMARKAPI_BASICAUTH",
            "%%Forwarded"
          ]
        }
      }
    }
```

**Comment by gganssauge** [comment]
That doesn't change anything unfortunately.
The access log I showed displays the request headers received by my ssl termination proxy (took some effort to generate it BTW).
No with the change it shows an additional header, but the x-forward-proto header is till present:
`sslproxy     | nginx.1    | api.test.smartpilot.haufe.io 40.113.12.39 - api-user [20/Oct/2016:20:59:28 +0000] "GET /bookmarks-api/v1/user/gotti HTTP/1.0" 200 275 "-" "curl/7.47.0" - "JS summary\x0A\x0AMethod: GET\x0AHTTP version: 1.0\x0AHost: api.test.smartpilot.haufe.io\x0ARemote Address: 40.113.12.39\x0AURI: /bookmarks-api/v1/user/gotti\x0AHeaders:\x0A  header 'X-Real-IP' is '91.66.22.165'\x0A  header 'X-Forwarded-For' is '91.66.22.165, 91.66.22.165'\x0A  header 'X-Forwarded-Proto' is 'http'\x0A  header 'Host' is 'api.test.smartpilot.haufe.io'\x0A  header 'Connection' is 'close'\x0A  header 'User-Agent' is 'curl/7.47.0'\x0A  header 'Accept' is 'application/hal+json'\x0A  header 'X-Forwarded-Port' is '443'\x0A  header 'X-Consumer-ID' is '4a5222e2-4bf6-4031-b8ca-bfa8f40d707c'\x0A  header 'X-Consumer-Custom-ID' is 'e3f6ac2bbefc2331bfc7b839ab7c55a8d8df3251'\x0A  header 'X-Consumer-Username' is 'test$bookmarkapi'\x0A  header 'X-Consumer-Groups' is 'bookmarkapi'\x0A  header 'Authorization' is 'Basic YXBpLXVzZXI6YVpwdGFmM0luU3RBQU1hb21nTzA='\x0A  header 'Forwarded' is 'host=api.test.smartpilot.haufe.io;port=443;proto=https;prefix=/bookmarks-api/v1'\x0A  header 'Correlation-Id' is '92b8c934-1cb4-4916-ba04-9ceec5413990'\x0AArgs:\x0A"`

**Comment by gganssauge** [comment]
Found a workaround:

The nginx configuration in `/usr/local/kong/nginx..conf` hardcodes the X-Forwared-Proto header to $scheme and $scheme seems to be http. When you replace
`proxy_set_header X-Forwarded-Proto $scheme;`
by
`proxy_set_header X-Forwarded-Proto https;`
everything works ok.

**Comment by DonMartin76** [comment]
OK, thanks. This has changed for Kong 0.9.x, and thus has to be revisited as soon as wicked 0.10.0 is released.

**Comment by DonMartin76** [comment]
Related to #16

**Comment by DonMartin76** [comment]
This is a Kong bug: https://github.com/Mashape/kong/issues/1823

**Comment by DonMartin76** [comment]
This is fixed via Kong 0.11.0 in wicked 0.12.0.
